### DISCUSSION TITLE
Allow seeing reason of version resolution failure

### DISCUSSION THREAD

**Discussion by Nicd** [gleam-lang/gleam#3635] [Ideas & suggestions]
In #3002 the reason for version resolution failures was changed to a more simplified version. Example:

```
➜  aurinko git:(trunk) ✗ gleam deps download
  Resolving versions
error: Dependency resolution failed

An error occurred while determining what dependency packages and
versions should be downloaded.
The error from the version resolver library was:

Unable to find compatible versions for the version constraints in your
gleam.toml. The conflicting packages are:

- aurinko
- mist
- wisp
```

When you have many packages that have many of their own dependencies, finding out the reason for the failure is not trivial even if it tells you which packages are involved. It would be useful to have a flag `--detailed-errors` or something like that to allow seeing the admittedly complicated error message that originally existed. I feel that it would be better than nothing. Currently I'm going through Hex.pm, trying to deduce the dependency ranges in my head, but it's difficult as I don't even know which dependency the failure is about.

**Comment by lpil** [comment]
When you say "see the reason" what do you mean? What would it look like?

**Comment by Nicd** [comment]
I would prefer being able to see this, even if it's difficult to read:

```
Because lustre_dev_tools [ 0.0.0, 1.0.1 [  [ 1.0.2, 1.1.0 [  [ 1.1.6, 1.2.0
[  [ 1.2.1, ∞ [ depends on glint 0.18.0 <= v < 1.0.0 and lustre_dev_tools
1.0.1 depends on glint 0.18.0 <= v < 1.0.0, lustre_dev_tools [ 0.0.0, 1.1.0
[  [ 1.1.6, 1.2.0 [  [ 1.2.1, ∞ [ depends on glint 0.18.0 <= v < 1.0.0.
And because lustre_dev_tools 1.1.0 depends on glint 0.18.0 <= v < 1.0.0,
lustre_dev_tools [ 0.0.0, 1.1.1 [  [ 1.1.6, 1.2.0 [  [ 1.2.1, ∞ [ depends
on glint 0.18.0 <= v < 1.0.0.
And because lustre_dev_tools 1.1.1 depends on glint 0.18.0 <= v <
1.0.0 and lustre_dev_tools 1.1.2 depends on glint 0.18.0 <= v < 1.0.0,
lustre_dev_tools [ 0.0.0, 1.1.3 [  [ 1.1.6, 1.2.0 [  [ 1.2.1, ∞ [ depends
on glint 0.18.0 <= v < 1.0.0.
And because lustre_dev_tools 1.1.3 depends on glint 0.18.0 <= v <
1.0.0 and lustre_dev_tools 1.1.4 depends on glint 0.18.0 <= v < 1.0.0,
lustre_dev_tools [ 0.0.0, 1.1.5 [  [ 1.1.6, 1.2.0 [  [ 1.2.1, ∞ [ depends
on glint 0.18.0 <= v < 1.0.0.
And because lustre_dev_tools 1.1.5 depends on glint 0.18.0 <= v <
1.0.0 and lustre_dev_tools 1.2.0 depends on glint 0.18.0 <= v < 1.0.0,
lustre_dev_tools depends on glint 0.18.0 <= v < 1.0.0.
And because buggy_error 0.0.0 depends on lustre_dev_tools and buggy_error
0.0.0 depends on glint 0.14.0, buggy_error 0.0.0 is forbidden.
```

Sure it's verbose and not nice to look at, but in the end when you read it carefully, you can get to the bottom of the problem faster than going through Hex, looking at the dependencies for all the packages, and trying to do the math yourself.

I can appreciate not showing this to a user as the first possible error, but as an advanced user I would like the option to see it.

**Reply by lpil** [reply]
This format is not good enough, it would need to be better.

**Reply by giacomocavalieri** [reply]
In this particular case I'd expect to see an error that looks something like this (supposing `buggy_error` is the name of the Gleam project you're compiling):
```text
error: Dependency resolution failed

I couldn't find a version of `lustre_dev_tools` compatible with your project's constraints:
- all versions of `lustre_dev_tools` require `glint` to be `0.18.0 <= and < 1.0.0`
- but your project requires `glint` to be `= 0.14.0`
```

The example is arguably a really simple case (no version of a package you're adding satisfies your project's `gleam.toml` direct constraints), but I think we could approach this gradually and start with simpler cases like this one and default to the less detailed error message in case it gets more unwieldy.

How would this be as an error message?

**Reply by lpil** [reply]
It sounds like you're proposing adding a second version resolution algorithm alongside pubgrub?

**Reply by giacomocavalieri** [reply]
No, it would be a different way of showing that big error message coming from pubgrub. It's all info you should be able to get from the error data type it returns.

**Reply by lpil** [reply]
If you can print their error nicely that sounds great! I don't think the text you've given there can be generated from the error though, it doesn't contain that information.

**Reply by giacomocavalieri** [reply]
Mmh maybe I was a bit too generic with the "all versions of `lustre_dev_tools`" part but it would be totally doable I think, it's all info I deduced from that error message:
```txt
Because lustre_dev_tools [ 0.0.0, 1.0.1 [  [ 1.0.2, 1.1.0 [  [ 1.1.6, 1.2.0
[  [ 1.2.1, ∞ [ depends on glint 0.18.0 <= v < 1.0.0 and lustre_dev_tools
1.0.1 depends on glint 0.18.0 <= v < 1.0.0, lustre_dev_tools [ 0.0.0, 1.1.0
[  [ 1.1.6, 1.2.0 [  [ 1.2.1, ∞ [ depends on glint 0.18.0 <= v < 1.0.0.
And because lustre_dev_tools 1.1.0 depends on glint 0.18.0 <= v < 1.0.0,
lustre_dev_tools [ 0.0.0, 1.1.1 [  [ 1.1.6, 1.2.0 [  [ 1.2.1, ∞ [ depends
on glint 0.18.0 <= v < 1.0.0.
And because lustre_dev_tools 1.1.1 depends on glint 0.18.0 <= v <
1.0.0 and lustre_dev_tools 1.1.2 depends on glint 0.18.0 <= v < 1.0.0,
lustre_dev_tools [ 0.0.0, 1.1.3 [  [ 1.1.6, 1.2.0 [  [ 1.2.1, ∞ [ depends
on glint 0.18.0 <= v < 1.0.0.
And because lustre_dev_tools 1.1.3 depends on glint 0.18.0 <= v <
1.0.0 and lustre_dev_tools 1.1.4 depends on glint 0.18.0 <= v < 1.0.0,
lustre_dev_tools [ 0.0.0, 1.1.5 [  [ 1.1.6, 1.2.0 [  [ 1.2.1, ∞ [ depends
on glint 0.18.0 <= v < 1.0.0.
And because lustre_dev_tools 1.1.5 depends on glint 0.18.0 <= v <
1.0.0 and lustre_dev_tools 1.2.0 depends on glint 0.18.0 <= v < 1.0.0,
// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
// All of this gibberish is just an explanation for this last line below


lustre_dev_tools depends on glint 0.18.0 <= v < 1.0.0.
// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
// All versions of `lustre_dev_tools` require `glint` to be `>= 0.18.0 and <= 1.0.0`


And because buggy_error 0.0.0 depends on lustre_dev_tools and buggy_error
0.0.0 depends on glint 0.14.0, buggy_error 0.0.0 is forbidden.
// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
// This is saying your project requires `glint` to be `= 0.14.0`
```

So it's all data in the error type, if the default pretty printer managed to display this error we should be able to condense it as I showed (I have no idea how hard that could be, but my guess is a lot! 😁)

**Reply by Nicd** [reply]
> This format is not good enough, it would need to be better.

I feel like this is letting perfect be the enemy of good. Or let's say letting good be the enemy of "better than nothing". While the error format is a real mouthful, you _can_ read it and I would have been glad to have it. I don't care if the flag is `--include-lousy-error-details`, at least it would exist.

**Reply by lpil** [reply]
I'm not asking for perfect, I'm asking for not-terrible. The bar is low and I would accept any half-decent format.

**Comment by giacomocavalieri** [comment]
A little update after asking for feedback on Discord as well! A nice looking output that I'm working on could look something like this:
```txt
error: Dependency resolution failed

An error occurred while determining what dependency packages and
versions should be downloaded.
The error from the version resolver library was:

- your package requires `two` to be `1.0.0`
- it also requires `wisp` to be `1.1.0`
- and `wisp` requires `gleam_json` to be `>= 0.6.0 and < 2.0.0`
- `two` requires `three` to be `1.0.0`
- and `three` requires `gleam_json` to be `>= 2.0.0 and < 3.0.0`

That means there's no version of `gleam_json` compatible with
both `wisp` and `three` together.
```

The last sentence was proposed by Hayleigh with the following reasoning:

> I think its not fininshed. you've laid the facts but not explained the conclusion (that wisp and three are incompatible)

And I agree that it would be better if after laying out all the facts the tool could help you reach the conclusion that there's no `gleam_json` version that satisfies all those constraints. I'm not sure how easy it would be to do but it's definitely worth looking into it!

**Reply by lpil** [reply]
Looks good to me! Can always improve it later
